fish-rust/src/env/env_ffi.rs

```diff
@@ -0,0 +1,368 @@
+        let statuses = Statuses {
+            status,
+            kill_signal: if kill_signal == 0 {
+                None
+            } else {
+                Some(Signal::new(kill_signal))
+            },
+            pipestatus: pipestatus.as_slice().to_vec(),
+        };
+        self.0.set_last_statuses(statuses)
+    }
+
+    fn set_pwd_from_getcwd(&self) {
+        self.0.set_pwd_from_getcwd()
+    }
+
+    fn set(&self, name: &CxxWString, flags: u16, vals: &wcstring_list_ffi_t) -> EnvStackSetResult {
+        let mode = EnvMode::from_bits(flags).expect("Invalid mode bits");
+        self.0.set(name.as_wstr(), mode, vals.from_ffi())
+    }
+
+    fn remove(&self, name: &CxxWString, flags: u16) -> EnvStackSetResult {
+        let mode = EnvMode::from_bits(flags).expect("Invalid mode bits");
+        self.0.remove(name.as_wstr(), mode)
+    }
+
+    fn export_array(&self) -> *mut c_char {
+        Box::into_raw(Box::new(OwningNullTerminatedArrayRefFFI(
+            self.0.export_array(),
+        )))
+        .cast()
+    }
+
+    fn snapshot(&self) -> Box<EnvDynFFI> {
+        Box::new(EnvDynFFI(self.0.snapshot()))
+    }
+
+    fn universal_sync(
+        self: &EnvStackRefFFI,
+        always: bool,
+        mut out_events: Pin<&mut event_list_ffi_t>,
+    ) {
+        let events: Vec<Box<Event>> = self.0.universal_sync(always);
+        for event in events {
+            out_events.as_mut().push(Box::into_raw(event).cast());
+        }
+    }
+}
+
+impl Statuses {
+    fn get_status_ffi(&self) -> i32 {
+        self.status
+    }
+
+    fn get_pipestatus_ffi(&self) -> &Vec<i32> {
+        &self.pipestatus
+    }
+
+    fn get_kill_signal_ffi(&self) -> i32 {
+        match self.kill_signal {
+            Some(sig) => sig.code(),
+            None => 0,
+        }
+    }
+}
+
+fn env_get_globals_ffi() -> Box<EnvStackRefFFI> {
+    Box::new(EnvStackRefFFI(EnvStack::globals().clone()))
+}
+
+fn env_get_principal_ffi() -> Box<EnvStackRefFFI> {
+    Box::new(EnvStackRefFFI(EnvStack::principal().clone()))
+}
+
+// We have to implement these directly to make cxx happy, even though they're implemented in the EnvironmentFFI trait.
+impl EnvNull {
+    pub fn getf_ffi(&self, name: &CxxWString, mode: u16) -> *mut EnvVar {
+        EnvironmentFFI::getf_ffi(self, name, mode)
+    }
+    pub fn get_names_ffi(&self, flags: u16, out: Pin<&mut wcstring_list_ffi_t>) {
+        EnvironmentFFI::get_names_ffi(self, flags, out)
+    }
+}
+
+trait EnvironmentFFI: Environment {
+    /// FFI helper.
+    /// This returns either null, or the result of Box.into_raw().
+    /// This is a workaround for the difficulty of passing an Option through FFI.
+    fn getf_ffi(&self, name: &CxxWString, mode: u16) -> *mut EnvVar {
+        match self.getf(
+            name.as_wstr(),
+            EnvMode::from_bits(mode).expect("Invalid mode bits"),
+        ) {
+            None => std::ptr::null_mut(),
+            Some(var) => Box::into_raw(Box::new(var)),
```

Comment:
returning `*mut Box<EnvVar>` would probably make things simpler

Reply:
I am not sure this is possible without constructing a `Box<Box<EnvVar>>`